src/mlpack/core/data/string_encoding.hpp

```diff
@@ -123,11 +124,12 @@ class StringEncoding
   *
   */
   template<typename TokenizerType>
-  typename std::enable_if<std::is_same< EncodingPolicy,
-      data::DictionaryEncoding>::value, void>::type Encode(const
+  void Encode(const
       std::vector<std::string>& input,
       std::vector<std::vector<size_t>>& output,
-      TokenizerType tokenizer);
+      TokenizerType tokenizer,
+      typename std::enable_if<PolicyTraits<EncodingPolicy>::
```

Comment:
I am not sure that it would work. The `Encode()` function does have a template argument (`tokenizer`) but the condition in `std::enable_if` statement doesn't depend on it. Actually, the condition and the policy are substituted at the class level. In this case SFINAE may not work.

Probably we should move the policy template argument to the `Encode()` function (i.e. outside of the class). I mean
the following:

```
class StringEncoding {
  template<typename Tokenizer, typename EncodingPolicy>
  void Encode(input, output,..., EncodingPolicy policy);
};
```
However I am not sure. Actually it's easy to test. Try to set `outputWithNoPadding=false` for `DictionaryPolicy` and invoke the `Encode()` function with a `vector<vector<size_t>>` as the output.

Reply:
Throws following error : 

```
/home/jeffin/mlpack/src/mlpack/core/data/string_encoding_impl.hpp:123:6: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
 void StringEncoding<EncodingPolicy>::Encode(
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
> Probably we should move the policy template argument to the `Encode()` function (i.e. outside of the class). I mean
> the following:
> 
> ```
> class StringEncoding {
>   template<typename Tokenizer, typename EncodingPolicy>
>   void Encode(input, output,..., EncodingPolicy policy);
> };
> ```

You mean to avoid template argument for class and add template argument for each of the function ?

Reply:
I presume the error still persists if you execute the `Encode()` function with `arma::mat` as the output.
> You mean to avoid template argument for class and add template argument for each of the function ?

Right, `EncodingPolicy` should be a template argument of the `Encode()` function rather than a template argument of the class.

Reply:
Done, Can you check once.